joomla/joomla-cms
# [5.2] Fix custom field attribute 'showon' having no effect in the frontend (Take 2)

Pull Request for Issue #40762 and follow up of #39051.

### Summary of Changes

The use of the parameter 'showon' has been introduced in Joomla 4.3 for custom fields.
However, a major issue is that field values are saved, even when hidden, which leads to values that are wrongly stored in the database and resulting to values that are rendered incorrectly in the front end. 

It is a redo of PR https://github.com/joomla/joomla-cms/pull/41104

### Testing Instructions

In the 'Content' context:
Create a field of type 'text' named 'Conditional Text'.
Create a field of type 'list' named 'Box'. Add options (value1, value2 and value3).
Create a field of type 'list' named 'Square'. Add options (value1, value2 and value3).

Contrary to PR #41104, you now need to test the fields on 'save' when in an article.
No matter the combination tested with showon, always add text to 'Conditional Text'.
For each combination:
- use a combination that shows the 'Conditional Text' field. Fill the value and 'save'. All values should be there after saving.
- use a combination that hides the 'Conditional Text' field (but fill the field anyway, like if you did, as user, decide and change your mind on the values selected). Save. Try a combination that would show the text field. The value of 'Conditional Text' should be empty.

Note: all combinations have already been tested in PR #41104 and the code for matching fields has not changed, so just pick up a single combination for the tests, no need to try them all anymore.

Values of the 'showon' attribute you can try out, as examples:
'Conditional Text' will show if:
`box:value1` => Box has value1
`box!:value1` => Box does not have value1
`box!:` => Box has a value that is not empty (in this case, you could add an empty option to that 'Box' field)
`box:value1[OR]square:value1` => Box has value1 or Square has value1
`box:value1[AND]square:value1` => Box and Square must have value1
`box:value1[OR]square!:value1` => Box has value1 or Square does not have value1
`box:value1[AND]square!:value1` => Box has value1 and Square has a value other than value1
`box:value1[AND]square:value1,value2` => Box has value1 and Square has value1 or value2
`box:value1[AND]square!:value1,value2` => Box has value1 and Square has value3

Note: this PR does not validate the 'showon' attribute field, nor handles the issue where a hidden field has been marked as 'required'. The documentation should properly explain that a field that can be hidden should never been required.
 
Test wrong entries in the 'showon' parameter with values such as `tada`, `box::`, `box:!`, `box:3:21`, `[AND]box:value3[OR]tada:2:3`... 
The combinations should be ignored and no error should be shown on the form.

To test subforms, make 'Conditional Text', 'Box' and 'Square' part of a subform field.
Note that in subforms, the 'showon' parameter must use a different syntax, since the field names, in a subform, show as 'field[id]'. For instance, if 'Box' has id 15, the name of 'box', in the subform, will be 'field15'.
So you can try testing with 'showon' values such as  `field15:value1`.
Test the values while saving an article, either in back or frontend.

To test custom fields sent through emails, you will need to create fields in the 'Contact' - 'Mail ' context.
Create a field of type 'text' named 'Conditional Subject'.
Create a field of type 'list' named 'Sub'. Add options (value1, value2 and value3).
Make sure both fields have proper permissions to be editable in the contact form.
Set a 'showon' value for the text field, for example: `sub:value1`.
Now, when you go to the front-end and display the form for a contact, you should be able to switch between values of 'Sub' and check that the text field shows or not. When the email is sent, the content of that email should reflect the choices made.

### Actual result BEFORE applying this Pull Request

All fields show in the front-end, no matter the values that have been set in the 'showon' parameter.
Same goes for fields that will show in emails sent through the contact component.

### Expected result AFTER applying this Pull Request

Fields are not saved if hidden and do not populate the database.
The values show properly in the frontend.

### Notes to developers:

- why add the code to match the showon parameter in FieldsHelper? So it can be made available globally and used in overrides and third-party extensions easily.

- I created onCustomFieldsBeforeSave for custom fields, an event that did not exist in the core before. 

### Link to documentations
Please select:
